### PR TITLE
support Azure File Service volume

### PR DESCRIPTION
**Pull request by rootfs**
[Azure File Storage](https://azure.microsoft.com/en-us/services/storage/files/) provides a SMB 3.0 file share. 

This implementation enables containers running on Azure to access file share located in the same Azure region.

### CODE REVIEW

File: examples/azure_file/azure.yaml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  - image: kubernetes/pause
+    name: azure
+    volumeMounts:
+      - name: azure
+        mountPath: /mnt/azure
+ volumes:
+      - name: azure
+        azureFile:
+          accountName: account_name
+          accountKey: account_key
```
Comment (swagiaal) [inline examples/azure_file/azure.yaml]: Hmm.. other implementations `gce_pd` and `aws_ebs` for example have kept account information outside the volume definition. Is it possible to do that with Azure ? Perhaps there is a way to authorize the instance through the Azure. I think that would be a better long term strategy especially after the implementation of the Attach/Detach controller.

File: pkg/api/types.go [hunk 2]
```diff
@@ -617,6 +621,19 @@ type DownwardAPIVolumeFile struct {
 	FieldRef ObjectFieldSelector `json:"fieldRef"`
 }
 
+// AzureFile represents an Azure File Service mount on the host that shares a pod's lifetime
```
Comment (swagiaal) [inline pkg/api/types.go]: s/AzureFile/AzureFileVolumeSource

File: pkg/api/v1/types.go [hunk 3]
```diff
@@ -737,6 +741,19 @@ type FCVolumeSource struct {
 	ReadOnly bool `json:"readOnly,omitempty"`
 }
 
+// AzureFile represents an Azure File Service mount on the host that shares a pod's lifetime
```
Comment (swagiaal) [inline pkg/api/v1/types.go]: s/AzureFile/AzureFileVolumeSource

File: pkg/volume/azure_file/azure_file.go [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+func (plugin *azureFilePlugin) Init(host volume.VolumeHost) {
+	plugin.host = host
+}
+
+func (plugin *azureFilePlugin) Name() string {
+	return azureFilePluginName
+}
+
+func (plugin *azureFilePlugin) CanSupport(spec *volume.Spec) bool {
+	//TODO: check if mount.cifs is there
```
Comment (swagiaal) [inline pkg/volume/azure_file/azure_file.go]: You should be able to use this #16950 soon

File: pkg/volume/azure_file/azure_file.go [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+	if !notMnt {
+		return nil
+	}
+	if err = b.util.SetupAzureFileSvc(b.accountName, b.accountKey, b.shareName); err != nil {
+		return err
+	}
+	os.MkdirAll(dir, 0750)
+	source := fmt.Sprintf("//%s.file.core.windows.net/%s", b.accountName, b.shareName)
+	// parameters suggested by https://azure.microsoft.com/en-us/documentation/articles/storage-how-to-use-files-linux/
+	options := []string{fmt.Sprintf("vers=3.0,username=%s,password=%s,dir_mode=0777,file_mode=0777", b.accountName, b.accountKey)}
```
Comment (swagiaal) [inline pkg/volume/azure_file/azure_file.go]: Hmm if accountKey is your password then revealing it in the pod definition might be problematic.

File: pkg/api/types.go (hunk 2 above)
Comment (rootfs) [inline pkg/api/types.go]: `AzureFile` is the variable name, the type is `AzureFileVolumeSource`

**Comment by rootfs** [comment]
blocked by https://github.com/kubernetes/kubernetes/issues/17433

**Comment by bgrant0607** [comment]
cc @thockin

**Comment by pmorie** [comment]
@rootfs I think this will be a lot easier to review if you split out into a commit for godeps and a commit for the kube changes.

**Comment by rootfs** [comment]
all green
@saad-ali @kubernetes/rh-storage

File: examples/azure_file/azure.yaml (hunk 1 above)
Comment (markturansky) [inline examples/azure_file/azure.yaml]: I agree with @swagiaal.  We should not embed the account info directly into the volume.

Why not a CloudProvider like aws/gce/cinder?

Comment (rootfs) [inline examples/azure_file/azure.yaml]: this is updated.

Comment (rootfs) [inline examples/azure_file/azure.yaml]: azure has separate storage and compute accounts. In AWS and GCE we use instance because we provision block storage to instance. Azure File storage can be provisioned without an instance. You can even access file share off Azure cloud.

File: pkg/api/v1/types.go [hunk 6]
```diff
@@ -311,6 +313,8 @@ type PersistentVolumeSource struct {
 	FC *FCVolumeSource `json:"fc,omitempty"`
 	// Flocker represents a Flocker volume attached to a kubelet's host machine and exposed to the pod for its usage. This depends on the Flocker control service being running
 	Flocker *FlockerVolumeSource `json:"flocker,omitempty"`
+	// AzureFile represents an Azure File Service mount on the host that shares a pod's lifetime
+	AzureFile *AzureFileVolumeSource `json:"azureFile,omitempty"`
```
Comment (pmorie) [inline pkg/api/v1/types.go]: This seems like it should be called `AzureFileService` and the volume source should be called `AzureFileServiceVolumeSource`

**Comment by rootfs** [comment]
@saad-ali PTAL, thanks!

**Comment by rootfs** [comment]
CI failed on `godep: restore, during download dep "github.com/docker/docker/pkg/timeutils": exit status 1`. Looks a broken godep. Docker pkg no longer has the timeutil any more, per [this commit](https://github.com/docker/docker/commit/27220ecc6b1eedf650ca9cf94965cb0dc2054efd) @dchen1107 can you help?

**Comment by rootfs** [comment]
blocked by #18833

**Comment by saad-ali** [comment]
I'm going to strongly suggest holding off adding additional volume plugins (unless there is a pressing need for a particular one) until https://github.com/kubernetes/kubernetes/issues/18333 is resolved. Therefore, I'm going to hold of code reviewing this until then. Let me know if you disagree.

**Comment by colemickens** [comment]
Is #18333 expected to be resolved and implemented before the next release? If not, can this be looked at more immediately? It seems there haven't been new comments on #18333 in a while and there is another issue opened for another storage plugin.

As long as this and the Flex storage driver are not available, it seems the next release will still have a bad story for cloud providers who don't already have storage volume plugins in the code base.

**Comment by thockin** [comment]
I spoke to Saad today.  I don't have any real reason to block new volume
plugins so long as they meet the quality bar and have an active
maintainer.  We decided that we're keeping plugins more or less as-is, so
one more isn't going to be a problem.

**Comment by saad-ali** [comment]
@colemickens https://github.com/kubernetes/kubernetes/issues/18333 will not make it in to v1.2, but should be part of 1.3. The Flex volume plugin is already merged and should be part of v1.2. Since it sounds like having a Azure volume plugin for v1.2 is important, I will prioritize reviewing this, and try to get it into v1.2 (that will depend on how much alteration is required, since the code freeze deadline is fast approaching).

**Comment by rootfs** [comment]
@saad-ali @thockin cool, I'll rebase it. Thanks.

**Comment by rootfs** [comment]
@saad-ali rebased, PTAL

File: pkg/volume/azure_file/doc.go [hunk 7]
```diff
@@ -0,0 +1,19 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
```
Comment (saad-ali) [inline pkg/volume/azure_file/doc.go]: Update copyright year for all new files.

File: pkg/volume/azure_file/azure_util.go [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+	}
+
+	//create azure client instance
+	client, err := azure.NewBasicClient(accountName, accountKey)
+	if err != nil {
+		return "", fmt.Errorf("failed to create azure storage client: %v", err)
+	}
+	svc := client.GetFileService()
+	//create azure file share if not exists yet
+	if _, err := svc.CreateShareIfNotExists(shareName); err != nil {
```
Comment (saad-ali) [inline pkg/volume/azure_file/azure_util.go]: Attach/Mount code paths should not create file share if it doesn't exist (should error out). Only provisioner code paths should.

File: pkg/api/types.go [hunk 9]
```diff
@@ -670,6 +674,19 @@ type DownwardAPIVolumeFile struct {
...
+type AzureFileVolumeSource struct {
+	// Azure File Service Account Name
+	AccountName string `json:"accountName"`
```
Comment (saad-ali) [inline pkg/api/types.go]: Put the `AccountName` inside secret along with the account key.

File: pkg/api/types.go [hunk 10]
```diff
@@ -670,6 +674,19 @@ type DownwardAPIVolumeFile struct {
...
+	// the name of Azure File Service Account Key ending in ==
+	KeyName string `json:"keyName"`
```
Comment (saad-ali) [inline pkg/api/types.go]: `KeyName` is a very generic and confusing name, and the description is misleading. This field contains the name of the *secret* that contains the credentials for this Azure volume. Name it something like `SecretName` and clarify the comment to indicate it refers to a secret which contains the Azure account name/key (not the actual Azure account key).

File: examples/azure_file/azure-key.yaml [hunk 11]
```diff
@@ -0,0 +1,7 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: azure-key
+type: Opague
+data:
+  key: Zm9vIC1uCg==
```
Comment (saad-ali) [inline examples/azure_file/azure-key.yaml]: For clarity, change `key` to `azureAccountKey`.
Move the account name here as `azureAccountName` (from API).
Document in `examples/azure_file/README.md` that the secret for Azure must be defined like this.

File: examples/azure_file/README.md [hunk 12]
```diff
@@ -0,0 +1,63 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
```
Comment (saad-ali) [inline examples/azure_file/README.md]: Also update `docs/user-guide/volumes.md`.

File: pkg/volume/azure_file/azure_util.go [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+	kubeClient := host.GetKubeClient()
+	if kubeClient == nil {
+		return "", fmt.Errorf("Cannot get kube client")
+	}
+
+	keys, err := kubeClient.Secrets(nameSpace).Get(keyName)
+	if err != nil {
+		return "", fmt.Errorf("Couldn't get account key %v/%v", nameSpace, keyName)
+	}
+	for _, data := range keys.Data {
```
Comment (saad-ali) [inline pkg/volume/azure_file/azure_util.go]: Explicitly look for `azureAccountKey` and `azureAccountName`

File: pkg/volume/azure_file/azure_file.go [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package azure_file
+
```
Comment (saad-ali) [inline pkg/volume/azure_file/azure_file.go]: Azure supports `CreateShare` and `DeleteShare` so please implement volume `Provisioner` and `Deleter` interfaces.

File: examples/azure_file/README.md (hunk 12 above)
Comment (rootfs) [inline examples/azure_file/README.md]: added

File: pkg/volume/azure_file/azure_file.go (hunk 14 above)
Comment (rootfs) [inline pkg/volume/azure_file/azure_file.go]: @saad-ali given that provisioning is WIP by @markturansky , and that Azure storage setup is different from AWS/GCE/OpenStack, I'd wait till provisioner is able to cope with these differences. 

Thought?

Comment (markturansky) [inline pkg/volume/azure_file/azure_file.go]: @saad-ali Provisioning Azure is not as straightforward as the other 3.   The one's we have now have a cloud provider and an API and they all work the same way.  Azure requires different authn/authz, needs user secrets, etc.

I think designing and implementing all that is out of scope for 1.2.  We'll need to design the next version of provisioning with this stuff in mind.

Comment (saad-ali) [inline pkg/volume/azure_file/azure_file.go]: Ok, let's leave dynamic provisioning out of this then.

Comment (rootfs) [inline pkg/volume/azure_file/azure_file.go]: cool, @saad-ali. Any more comments? I'll fix and rebase again.

**Comment by saad-ali** [comment]
> cool, @saad-ali. Any more comments? I'll fix and rebase again.

Please address the existing comments, then I'll take another look.

**Comment by rootfs** [comment]
@saad-ali feedback addressed and rebased.

**Comment by saad-ali** [comment]
Ping me when this is ready for review.

**Comment by rootfs** [comment]
@saad-ali PTAL, thanks.

File: pkg/volume/azure_file/azure_util.go [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+	}
+
+	//create azure client instance
+	client, err := azure.NewBasicClient(accountName, accountKey)
+	if err != nil {
+		return "", "", fmt.Errorf("failed to create azure storage client: %v", err)
+	}
+	svc := client.GetFileService()
+	//create azure file share if not exists yet
+	if _, err := svc.CreateShareIfNotExists(shareName); err != nil {
```
Comment (saad-ali) [inline pkg/volume/azure_file/azure_util.go]: Didn't see a response to this comment:
> Attach/Mount code paths should not create file share if it doesn't exist (should error out). Only provisioner code paths should.

File: examples/azure_file/secret/azure-secret.yaml [hunk 16]
```diff
@@ -0,0 +1,8 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: azure-secret
+type: Opague
```
Comment (saad-ali) [inline examples/azure_file/secret/azure-secret.yaml]: Opaque not Opague

File: pkg/api/types.go [hunk 17]
```diff
@@ -215,6 +215,8 @@ type VolumeSource struct {
 	DownwardAPI *DownwardAPIVolumeSource `json:"downwardAPI,omitempty"`
 	// FC represents a Fibre Channel resource that is attached to a kubelet's host machine and then exposed to the pod.
 	FC *FCVolumeSource `json:"fc,omitempty"`
+	// AzureFile represents an Azure File Service mount on the host that shares a pod's lifetime
```
Comment (saad-ali) [inline pkg/api/types.go]: `shares a pod's lifetime` -> I know this is the comment used on a bunch of existing plugins. But it doesn't really seem correct. The whole point of a these network attached storage volumes is that they persist beyond the lifecycle of a pod. You could argue that the mount itself exists for the duration of a pod lifetime.

In short, if you can come up with better wording for this, that would be great, if not, no biggie.

File: pkg/volume/azure_file/azure_util.go (hunk 15 above)
Comment (saad-ali) [inline pkg/volume/azure_file/azure_util.go]: And if this is removed, then you don't really need to create the file service, and if you don't need to create the file service you don't need to create the client, which means you don't need the Azure API at all and can remove it from GoDeps.

**Comment by rootfs** [comment]
@saad-ali Updated. PTAL. thanks

File: pkg/volume/azure_file/azure_util.go [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+)
+
+// Abstract interface to azure file operations.
+type azureUtil interface {
+	SetupAzureFileSvc(host volume.VolumeHost, nameSpace, secretName, shareName string) (string, string, error)
+}
+
+type azureSvc struct{}
+
+func (s *azureSvc) SetupAzureFileSvc(host volume.VolumeHost, nameSpace, secretName, shareName string) (string, string, error) {
```
Comment (saad-ali) [inline pkg/volume/azure_file/azure_util.go]: nit: Method name no longer accurate. Maybe `GetAzureCredentials`?

**Comment by saad-ali** [comment]
LGTM mod nit + rebase.

**Comment by rootfs** [comment]
@saad-ali updated and rebased. thanks.

**Comment by saad-ali** [comment]
LGTM

Thanks @rootfs
